File: 82_remove_duplicates_from_sorted_list_II/deleteDuplicates.py
class Solution(object):
    def deleteDuplicates(self, head):

        count = 1
        while head and head.next and:
            if head.val == head.next.val:
                head = head.next
                count += 1
        
        if count > 1:
            head = head.next
        
        tmp = head
        prev = None


        
        
        while tmp and tmp.next:
            if tmp.val == tmp.next.val:
                if tmp.next.next:
                    tmp.next = tmp.next.next
                else:
                    prev.next = None
            else:
                if prev == None:
                    head = tmp.next
                prev = tmp
                tmp = tmp.next


        
        return head

File: 82_remove_duplicates_from_sorted_list_II/deleteDuplicates.cpp
#include <iostream>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
	public:
    	ListNode* deleteDuplicates(ListNode* head) {
            if(!head) return NULL;
            ListNode* prev = new ListNode(-1);
            ListNode* dummy = new ListNode(head->val + 1);
            dummy -> next = head;
            prev -> next = dummy;
            ListNode* tmp = dummy;
            while(tmp && tmp -> next) {
                if(tmp->val == tmp->next->val) {
                    while(tmp && tmp->next && (tmp->val == tmp->next->val)) {
                        tmp = tmp -> next;
                    }
                    tmp = tmp -> next;
                    prev -> next = tmp;
                }
                else {
                    prev = prev -> next;
                    tmp = tmp -> next;
                }
            }

            return dummy -> next;
    	}
};
